omod/src/test/java/org/openmrs/module/idgen/rest/controller/AutoGenerationOptionControllerTest.java

```diff
@@ -0,0 +1,134 @@
+	}
+	
+	@Override
+	public long getAllCount() {
+		Integer allAutoGenerationOptions = 0;
+		for (PatientIdentifierType patientIdentifierType : Context.getPatientService().getAllPatientIdentifierTypes()) {
+			allAutoGenerationOptions += identifierSourceService.getAutoGenerationOptions(patientIdentifierType).size();
+		}
+		return allAutoGenerationOptions;
+	}
+	
+	@Test
+	public void shouldGetAnAutoGenerationOptionByPatientIdentifierType() throws Exception {
+		MockHttpServletRequest req = request(RequestMethod.GET, getURI() + "/" + getUuid());
+		SimpleObject result = deserialize(handle(req));
+		PatientIdentifierType patientIdentifier = Context.getPatientService().getPatientIdentifierTypeByUuid(getUuid());
+		AutoGenerationOption autoGenerationOption = identifierSourceService.getAutoGenerationOption(patientIdentifier);
+		assertEquals(autoGenerationOption.getIdentifierType().getUuid(),
+		    PropertyUtils.getProperty(result, "identifierType"));
+		assertEquals(autoGenerationOption.getLocation().getUuid(), PropertyUtils.getProperty(result, "location"));
+		assertEquals(autoGenerationOption.getSource().getUuid(), PropertyUtils.getProperty(result, "source"));
+	}
+	
+	@Test
+	public void shouldCreateNewAutoGenerationOption() throws Exception {
+		long originalCount = getAllCount();
+		boolean automaticGenerationEnabled = true;
+		boolean manualEntryEnabled = true;
+		SimpleObject autoGenerationOption = new SimpleObject();
+		autoGenerationOption.add("identifierType", getUuid());
+		autoGenerationOption.add("location", LOCATION);
+		autoGenerationOption.add("source", SOURCE);
+		autoGenerationOption.add("automaticGenerationEnabled", automaticGenerationEnabled);
+		autoGenerationOption.add("manualEntryEnabled", manualEntryEnabled);
+		String json = new ObjectMapper().writeValueAsString(autoGenerationOption);
+		MockHttpServletRequest req = request(RequestMethod.POST, getURI());
+		req.setContent(json.getBytes());
+		SimpleObject newAutoGenerationOption = deserialize(handle(req));
+		assertNotNull(PropertyUtils.getProperty(newAutoGenerationOption, "uuid"));
+		assertEquals(originalCount + 1, getAllCount());
+	}
+	
+	@Test
+	public void shouldEditAnAutoGenerationOption() throws Exception {
+		boolean automaticGenerationEnabled = false;
+		boolean manualEntryEnabled = false;
+		SimpleObject autoGenerationOption = new SimpleObject();
+		autoGenerationOption.add("location", NEW_LOCATION);
+		autoGenerationOption.add("source", NEW_SOURCE);
+		autoGenerationOption.add("manualEntryEnabled", manualEntryEnabled);
+		autoGenerationOption.add("automaticGenerationEnabled", automaticGenerationEnabled);
+		String json = new ObjectMapper().writeValueAsString(autoGenerationOption);
+		MockHttpServletRequest req = request(RequestMethod.POST, getURI() + "/" + getUuid());
+		req.setContent(json.getBytes());
+		handle(req);
+		PatientIdentifierType patientIdentifier = Context.getPatientService().getPatientIdentifierTypeByUuid(getUuid());
+		assertEquals(NEW_LOCATION,
```

Comment:
How sure are you that these were not the same values before the REST call was made?